File: tests/analytics/protein-distribution.test.ts
import { describe, test, expect } from 'vitest';
import {
	computeProteinDistribution,
	proteinPerMealThreshold
} from '$lib/analytics/protein-distribution';

function makeEntries(days: { date: string; meals: { mealType: string; protein: number }[] }[]) {
	return days.flatMap(({ date, meals }) =>
		meals.map(({ mealType, protein }) => ({ date, mealType, protein }))
	);
}

describe('computeProteinDistribution', () => {
	test('even distribution gives high score', () => {
		const entries = makeEntries([
			{
				date: '2024-01-01',
				meals: [
					{ mealType: 'Breakfast', protein: 33 },
					{ mealType: 'Lunch', protein: 33 },
					{ mealType: 'Dinner', protein: 34 }
				]
			}
		]);
		const result = computeProteinDistribution(entries);
		expect(result.score).toBeGreaterThan(90);
	});

	test('skewed distribution gives low score', () => {
		const entries = makeEntries([
			{
				date: '2024-01-01',
				meals: [
					{ mealType: 'Breakfast', protein: 10 },
					{ mealType: 'Lunch', protein: 10 },
					{ mealType: 'Dinner', protein: 80 }
				]
			}
		]);
		const result = computeProteinDistribution(entries);
		expect(result.score).toBeLessThan(50);
	});

	test('a single meal per day is the skewed pattern, not a perfect one', () => {
		const entries = makeEntries([
			{ date: '2024-01-01', meals: [{ mealType: 'Lunch', protein: 50 }] },
			{ date: '2024-01-02', meals: [{ mealType: 'Lunch', protein: 60 }] }
		]);
		const result = computeProteinDistribution(entries);
		// Padded to three feedings [50, 0, 0]: CV = √2 → score floors at 0.
		expect(result.score).toBe(0);
		expect(result.mealsPerDay).toBe(1);
	});

	test('two even meals score below three even meals', () => {
		const two = computeProteinDistribution(
			makeEntries([
				{
					date: '2024-01-01',
					meals: [
						{ mealType: 'Lunch', protein: 45 },
						{ mealType: 'Dinner', protein: 45 }
					]
				}
			])
		);
		const three = computeProteinDistribution(
			makeEntries([
				{
					date: '2024-01-01',
					meals: [
						{ mealType: 'Breakfast', protein: 30 },
						{ mealType: 'Lunch', protein: 30 },
						{ mealType: 'Dinner', protein: 30 }
					]
				}
			])
		);
		expect(three.score).toBe(100);
		expect(two.score).toBeLessThan(three.score);
		expect(two.score).toBeGreaterThan(0);
	});

	test('the per-meal bar scales with body weight', () => {
		expect(proteinPerMealThreshold(null)).toBe(20);
		expect(proteinPerMealThreshold(50)).toBe(20);
		expect(proteinPerMealThreshold(85)).toBeCloseTo(34, 9);
		const entries = makeEntries([
			{
				date: '2024-01-01',
				meals: [
					{ mealType: 'Lunch', protein: 25 },
					{ mealType: 'Dinner', protein: 40 }
				]
			}
		]);
		const result = computeProteinDistribution(entries, proteinPerMealThreshold(85));
		expect(result.threshold).toBeCloseTo(34, 9);
		expect(result.mealsBelowThreshold).toBe(1);
	});

	test('mealsBelowThreshold counts correctly', () => {
		const entries = makeEntries([
			{
				date: '2024-01-01',
				meals: [
					{ mealType: 'Breakfast', protein: 10 },
					{ mealType: 'Lunch', protein: 25 },
					{ mealType: 'Dinner', protein: 15 }
				]
			}
		]);
		const result = computeProteinDistribution(entries, 20);
		expect(result.mealsBelowThreshold).toBe(2);
	});

	test('returns zero values for empty input', () => {
		const result = computeProteinDistribution([]);
		expect(result.score).toBe(0);
		expect(result.sampleSize).toBe(0);
		expect(result.confidence).toBe('insufficient');
	});

	test('sampleSize is number of distinct days', () => {
		const entries = makeEntries([
			{ date: '2024-01-01', meals: [{ mealType: 'Breakfast', protein: 30 }] },
			{ date: '2024-01-02', meals: [{ mealType: 'Breakfast', protein: 30 }] },
			{ date: '2024-01-03', meals: [{ mealType: 'Breakfast', protein: 30 }] }
		]);
		const result = computeProteinDistribution(entries);
		expect(result.sampleSize).toBe(3);
	});

	test('aggregates multiple entries for same meal on same day', () => {
		const entries = [
			{ date: '2024-01-01', mealType: 'Lunch', protein: 20 },
			{ date: '2024-01-01', mealType: 'Lunch', protein: 10 },
			{ date: '2024-01-01', mealType: 'Dinner', protein: 30 }
		];
		const result = computeProteinDistribution(entries);
		expect(result.avgPerMeal).toBeCloseTo(30);
	});
});

File: mobile/shared/src/commonTest/kotlin/com/bissbilanz/analytics/ProteinDistributionTest.kt
package com.bissbilanz.analytics

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ProteinDistributionTest {
    @Test
    fun evenDistributionScoreNear100() {
        val entries =
            listOf(
                Triple("2024-01-01", "Breakfast", 30.0),
                Triple("2024-01-01", "Lunch", 30.0),
                Triple("2024-01-01", "Dinner", 30.0),
                Triple("2024-01-02", "Breakfast", 30.0),
                Triple("2024-01-02", "Lunch", 30.0),
                Triple("2024-01-02", "Dinner", 30.0),
            )
        val result = computeProteinDistribution(entries)
        assertTrue(result.score > 90.0, "Expected score near 100 for even distribution, got ${result.score}")
        assertEquals(30.0, result.avgPerMeal)
        assertEquals(3.0, result.mealsPerDay)
    }

    @Test
    fun highlyUnevenDistributionLowScore() {
        val entries =
            listOf(
                Triple("2024-01-01", "Breakfast", 5.0),
                Triple("2024-01-01", "Lunch", 5.0),
                Triple("2024-01-01", "Dinner", 100.0),
                Triple("2024-01-02", "Breakfast", 5.0),
                Triple("2024-01-02", "Lunch", 5.0),
                Triple("2024-01-02", "Dinner", 100.0),
            )
        val result = computeProteinDistribution(entries)
        assertTrue(result.score < 60.0, "Expected low score for uneven distribution, got ${result.score}")
    }

    @Test
    fun emptyEntriesReturnsZero() {
        val result = computeProteinDistribution(emptyList())
        assertEquals(0.0, result.score)
        assertEquals(0.0, result.avgPerMeal)
        assertEquals(0.0, result.mealsPerDay)
        assertEquals(0, result.totalMeals)
        assertEquals(ConfidenceLevel.INSUFFICIENT, result.confidence)
        assertEquals(0, result.sampleSize)
    }

    @Test
    fun singleMealPerDayHasPerfectCV() {
        val entries =
            listOf(
                Triple("2024-01-01", "Lunch", 40.0),
                Triple("2024-01-02", "Lunch", 50.0),
                Triple("2024-01-03", "Lunch", 45.0),
                Triple("2024-01-04", "Lunch", 42.0),
                Triple("2024-01-05", "Lunch", 48.0),
                Triple("2024-01-06", "Lunch", 44.0),
                Triple("2024-01-07", "Lunch", 46.0),
            )
        val result = computeProteinDistribution(entries)
        // Padded to three feedings [x, 0, 0]: the skewed pattern, not a perfect one.
        assertEquals(0.0, result.score, 1e-9)
        assertEquals(1.0, result.mealsPerDay)
    }

    @Test
    fun twoEvenMealsScoreBelowThreeEvenMeals() {
        val two = computeProteinDistribution(listOf(Triple("2024-01-01", "Lunch", 45.0), Triple("2024-01-01", "Dinner", 45.0)))
        val three =
            computeProteinDistribution(
                listOf(
                    Triple("2024-01-01", "Breakfast", 30.0),
                    Triple("2024-01-01", "Lunch", 30.0),
                    Triple("2024-01-01", "Dinner", 30.0),
                ),
            )
        assertEquals(100.0, three.score, 1e-9)
        assertTrue(two.score < three.score && two.score > 0.0)
    }

    @Test
    fun perMealThresholdScalesWithBodyWeight() {
        assertEquals(20.0, proteinPerMealThreshold(null))
        assertEquals(20.0, proteinPerMealThreshold(50.0))
        assertEquals(34.0, proteinPerMealThreshold(85.0), 1e-9)
        val result = computeProteinDistribution(listOf(Triple("2024-01-01", "Lunch", 25.0)), proteinPerMealThreshold(85.0))
        assertEquals(34.0, result.threshold, 1e-9)
        assertEquals(1, result.mealsBelowThreshold)
    }

    @Test
    fun mealsBelowThresholdCounted() {
        val entries =
            listOf(
                Triple("2024-01-01", "Breakfast", 10.0),
                Triple("2024-01-01", "Dinner", 30.0),
                Triple("2024-01-02", "Breakfast", 15.0),
                Triple("2024-01-02", "Dinner", 35.0),
                Triple("2024-01-03", "Breakfast", 8.0),
                Triple("2024-01-03", "Dinner", 40.0),
                Triple("2024-01-04", "Breakfast", 12.0),
                Triple("2024-01-04", "Dinner", 38.0),
            )
        val result = computeProteinDistribution(entries, threshold = 20.0)
        assertEquals(4, result.mealsBelowThreshold)
    }

    @Test
    fun singleEntryTotalHasOneMealPerDay() {
        val entries = listOf(Triple("2024-01-01", "lunch", 30.0))
        val result = computeProteinDistribution(entries)
        assertEquals(1, result.sampleSize)
        assertEquals(1.0, result.mealsPerDay)
        assertEquals(0.0, result.score, 1e-9)
    }

    @Test
    fun zeroProteinValuesCountedBelowThreshold() {
        val entries =
            listOf(
                Triple("2024-01-01", "breakfast", 0.0),
                Triple("2024-01-01", "lunch", 40.0),
                Triple("2024-01-01", "dinner", 0.0),
            )
        val result = computeProteinDistribution(entries)
        assertEquals(2, result.mealsBelowThreshold)
    }

    @Test
    fun customThresholdRespected() {
        val entries =
            listOf(
                Triple("2024-01-01", "breakfast", 15.0),
                Triple("2024-01-01", "lunch", 25.0),
            )
        val result10 = computeProteinDistribution(entries, threshold = 10.0)
        assertEquals(0, result10.mealsBelowThreshold)
        val result30 = computeProteinDistribution(entries, threshold = 30.0)
        assertEquals(2, result30.mealsBelowThreshold)
    }
}
